Firestore/core/test/firebase/firestore/local/counting_query_engine.h

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_TEST_FIREBASE_FIRESTORE_LOCAL_COUNTING_QUERY_ENGINE_H_
+#define FIRESTORE_CORE_TEST_FIREBASE_FIRESTORE_LOCAL_COUNTING_QUERY_ENGINE_H_
+
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "Firestore/core/src/firebase/firestore/core/query.h"
+#include "Firestore/core/src/firebase/firestore/local/mutation_queue.h"
+#include "Firestore/core/src/firebase/firestore/local/query_engine.h"
+#include "Firestore/core/src/firebase/firestore/local/remote_document_cache.h"
+#include "Firestore/core/src/firebase/firestore/model/document_key_set.h"
+#include "Firestore/core/src/firebase/firestore/model/document_map.h"
+#include "Firestore/core/src/firebase/firestore/model/snapshot_version.h"
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
+class LocalDocumentsView;
+class WrappedMutationQueue;
+class WrappedRemoteDocumentCache;
+
+/**
+ * A test-only QueryEngine that forwards all API calls and exposes the number of
+ * documents and mutations read.
+ */
+class CountingQueryEngine : public QueryEngine {
```

Comment:
Port of https://github.com/firebase/firebase-android-sdk/blob/2a2e789629a7dad13e1007b6cdec26c96e73defe/firebase-firestore/src/test/java/com/google/firebase/firestore/local/CountingQueryEngine.java#L36